[export] runtime asserts for while HOP subgraphs

pianpwk opened the pull request:
Differential Revision: D78431075

For #158366
- Calls runtime asserts pass for HOP subgraphs (in reenter_make_fx)
- For while_loop only (can be expanded), clones input tensors for subgraph tracing, so unbacked memos (item, nonzero, etc.) aren't reused


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela @zou3519 @ydwu4

ydwu4 commented on torch/_subclasses/fake_impls.py:
```diff
@@ -411,7 +413,9 @@ def repeat_interleave_tensor(fake_mode, func, repeats, output_size=None):
 @register_op_impl(torch.ops.aten._local_scalar_dense.default)
 def local_scalar_dense(fake_mode, func, arg):
-    if (r := arg.item_memo) is not None:
+    if (
```
is the item memo associated with the fake_tensor? 

Disabling item_memo may not be what we want  e.g. users might be doing something like:
```python
def body_fn(c):
  for _ in range(10):
    c_val = c.item()
    ...
```
In this case, we don't want to create an unbacked symint every iteration since they're essentially the same .item().

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1241,8 +1241,28 @@ def unspecialize_carried_inputs(tx, carry) -> VariableTracker:
-            new_operands_seq = [
-                unspecialize_carried_inputs(tx, carry) for carry in operands_seq
+            # clone inputs across subgraphs, to avoid unbacked memoization in fake prop
+            cond_operands_seq = [
```
this should be put under with discard_graph_changes(tx):

ydwu4 commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -319,12 +316,14 @@ def _unspecialize_carried_inputs(x):
+            )
+            return reenter_make_fx(fn)(*cloned_carried_inputs, *additional_inputs)
+
+        cond_graph = produce_graph(cond_fn)
```
put under with disable_proxy_modes_tracing():

pianpwk committed:
[export] runtime asserts for while HOP subgraphs (#158467)

Summary:

For #158366
- Calls runtime asserts pass for HOP subgraphs (in reenter_make_fx)
- For while_loop only (can be expanded), clones input tensors for subgraph tracing, so unbacked memos (item, nonzero, etc.) aren't reused

Test Plan:
test_export

Rollback Plan:

Reviewed By: ydwu4

Differential Revision: D78431075
